core/src/main/java/ai/timefold/solver/core/impl/bavet/uni/StaticDataUniNode.java

```diff
@@ -0,0 +1,180 @@
+package ai.timefold.solver.core.impl.bavet.uni;
+
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+
+import ai.timefold.solver.core.impl.bavet.NodeNetwork;
+import ai.timefold.solver.core.impl.bavet.common.AbstractNode;
+import ai.timefold.solver.core.impl.bavet.common.Propagator;
+import ai.timefold.solver.core.impl.bavet.common.StaticPropagationQueue;
+import ai.timefold.solver.core.impl.bavet.common.TupleSourceRoot;
+import ai.timefold.solver.core.impl.bavet.common.tuple.RecordingTupleNode;
+import ai.timefold.solver.core.impl.bavet.common.tuple.TupleLifecycle;
+import ai.timefold.solver.core.impl.bavet.common.tuple.TupleState;
+import ai.timefold.solver.core.impl.bavet.common.tuple.UniTuple;
+
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public final class StaticDataUniNode<A> extends AbstractNode implements TupleSourceRoot<Object>, TupleLifecycle<UniTuple<A>> {
+    private final StaticPropagationQueue<UniTuple<A>> propagationQueue;
+    private final Map<Object, List<UniTuple<A>>> tupleMap = new IdentityHashMap<>(1000);
+    private final int outputStoreSize;
+    private final NodeNetwork nodeNetwork;
+    private final RecordingTupleNode<UniTuple<A>> recordingTupleNode;
+    private final Class<?>[] sourceClasses;
+
+    public StaticDataUniNode(NodeNetwork nodeNetwork,
+            RecordingTupleNode<UniTuple<A>> recordingTupleNode,
+            int outputStoreSize,
+            TupleLifecycle<UniTuple<A>> nextNodesTupleLifecycle,
+            Class<?>[] sourceClasses) {
+        this.nodeNetwork = nodeNetwork;
+        this.outputStoreSize = outputStoreSize;
+        this.propagationQueue = new StaticPropagationQueue<>(nextNodesTupleLifecycle);
+        this.sourceClasses = sourceClasses;
+        this.recordingTupleNode = recordingTupleNode;
+    }
+
+    @Override
+    public boolean allowsInstancesOf(Class<?> clazz) {
+        for (var sourceClass : sourceClasses) {
+            if (sourceClass.isAssignableFrom(clazz)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Class<?>[] getSourceClasses() {
+        return sourceClasses;
+    }
+
+    @Override
+    public boolean supports(LifecycleOperation lifecycleOperation) {
+        return true;
+    }
+
+    @Override
+    public Propagator getPropagator() {
+        return propagationQueue;
+    }
+
+    @Override
+    public void insert(Object a) {
+        if (tupleMap.containsKey(a)) {
+            return;
+        }
+        invalidateCache();
+        tupleMap.put(a, new ArrayList<>());
+        insertIntoNodeNetwork(a);
+        recalculateTuples();
```

Comment:
Doing this on every insert/retract seems excessively expensive. There will be a large burst of inserts at the beginning; they ought to be cached and processed in a batch, probably before propagation starts.